File: 1.算法框架_labuladong/3.二叉搜索树/第一期/230寻找第 K 小的元素.ts
namespace LeetCode230_1 {
	function kthSmallest(root: TreeNode, k: number) {
		let res = 0
		let rank = 0
		const traverse = (root: TreeNode, k: number) => {
			if (!root) return
			traverse(root.left, k)
			rank++
			if (k == rank) {
				res = root.val
				return
			}
			traverse(root.right, k)
		}
		traverse(root, k)
		return res
	}
}

File: legacy/1.算法框架_labuladong/3.二叉搜索树/第一期/230寻找第 K 小的元素.java

class Solution {
	public int kthSmallest(TreeNode root, int k) {
		traverse(root, k);
		return res;
	}

	int res = 0;
	int rank = 0;

	void traverse(TreeNode root, int k) {
		if (root == null) {
			return;
		}
		traverse(root.left, k);
		rank++;
		if (k == rank) {
			res = root.val;
			return;
		}
		traverse(root.right, k);
	}
}
